modules/custom/az_migration/az_migration_remote/src/Plugin/migrate/process/MigrationRemoteFile.php

```diff
@@ -0,0 +1,278 @@
+        '%url' => $url,
+      ]);
+      return $original_uri;
+    }
+
+    // Get the fallback filename.
+    $filename = $row->get($this->configuration['default_filename']) ?? 'file';
+
+    $mimeTypes = new MimeTypes();
+    // Find the proper fallback file extension if possible.
+    $extensions = $mimeTypes->getExtensions($type);
+    $extension = '';
+    if (!empty($extensions)) {
+      $extension = reset($extensions);
+      $filename .= '.' . $extension;
+    }
+    else {
+      $this->logger->notice("Couldn't determine file extension for %url", [
+        '%url' => $url,
+      ]);
+      return;
+    }
+    $this->logger->debug("Recognized extension as %ext", [
+      '%ext' => $extension,
+    ]);
+    $disposition = $response->getHeader('Content-Disposition') ?? [];
+    $disposition = array_pop($disposition);
+    // See if we can determine the real remote filename from the disposition.
+    if (!empty($disposition)) {
+      preg_match(static::REQUEST_HEADER_FILENAME_REGEX, $disposition, $matches);
+      if (!empty($matches['filename'])) {
+        $filename = $matches['filename'];
+      }
+    }
+    $this->logger->debug("Determined filename as %filename", [
+      '%filename' => $filename,
+    ]);
+
+    $body = $response->getBody();
+
+    // Get directory path for new files.
+    $directory = $this->configuration['directory'] ?? 'public://';
+
+    // Sanitize our filename. Give other modules a change to weigh in.
+    $event = new FileUploadSanitizeNameEvent($filename, $extension);
+    $this->eventDispatcher->dispatch($event);
+    $filename = $event->getFilename();
+    $this->logger->debug("Sanitized to %filename", [
+      '%filename' => $filename,
+    ]);
+
+    // Check hash to determine whether to return this file or original.
+    $original_hash = NULL;
+    $uri = $original_uri;
+    if ($original_uri && file_exists($original_uri)) {
+      $original_hash = Crypt::hashBase64(file_get_contents($original_uri));
+    }
+    $hash = Crypt::hashBase64($body);
+    // Check if file has changed (or is new).
+    if ($hash !== $original_hash) {
+      // Create the actual disk file from the body.
+      try {
+        // Make sure our destination directory exists.
+        $this->fileSystem->prepareDirectory($directory, FileSystemInterface::CREATE_DIRECTORY | FileSystemInterface::MODIFY_PERMISSIONS);
+        $new_uri = $this->fileSystem->createFilename($filename, $directory);
+        $new_uri = $this->fileSystem->saveData($body, $new_uri, FileExists::Replace);
+        if (!empty($original_uri)) {
+          $this->logger->notice("Updating @original_uri to @new_uri because its hash has changed.", [
+            '@original_uri' => $original_uri,
+            '@new_uri' => $new_uri,
+          ]);
+        }
+        $uri = $new_uri;
+      }
+      catch (\Exception $e) {
+        $this->logger->notice("During migration of %url, @msg", [
+          '%url' => $url,
```

Comment:
```suggestion
          '%url' => $uri,
```
Is this supposed to be `$uri`?

Reply:
I'd not intended it to be `uri`, but after thinking about it I'm not sure if `url` is helpful either. Given the functions the `try{ }` is wrapping I guess the most likely problems are issues with creating the new file.

What would be most helpful? I wonder if maybe niether url/uri and instead the filename that it was trying to create, maybe... (which presumably is what failed)

Reply:
It looked to me like you were referencing the wrong variable here but I think I just wasn't following the code preceding this very well while reviewing.  I think it's probably OK as-is.